cyberpandas/ip_array.py

```diff
@@ -305,15 +305,39 @@ def isin(self, other):
             except ValueError:
                 networks.append(ipaddress.IPv6Network(net))
 
-        # TODO: perf
-        pyips = self.to_pyipaddress()
         mask = np.zeros(len(self), dtype='bool')
         for network in networks:
-            for i, ip in enumerate(pyips):
-                if ip in network:
-                    mask[i] = True
+            mask |= self._isin_network(network)
         return mask
 
+    def _isin_network(self, other):
+        # type: (Union[ipaddress.IPv4Network,ipaddress.IPv6Network]) -> ndarray
+        # https://stackoverflow.com/a/1004527/1889400
+        net_lo = int(other.network_address)
+        net_hi = int(other.broadcast_address)
+
+        if isinstance(other, ipaddress.IPv4Network):
+            # We know we only need the lower 64 bits
+            return (self.data['hi'] == 0) & ((net_lo <= self.data['lo']) &
+                                             (self.data['lo'] <= net_hi))
+        else:
+            # Okkkkkkkkkkkkkk, let's carefully watch overflow here.
+            if net_lo <= _U8_MAX and net_hi <= _U8_MAX:
+                # basically same as above?
+                return (self.data['hi'] == 0) & ((net_lo <= self.data['lo']) &
+                                                 (self.data['lo'] <= net_hi))
+            elif net_lo <= _U8_MAX:
+                # but net_hi is larger... Is this possible?
+                raise NotImplementedError("TODO")
+            else:
+                # The upper 64 bits match when hi_mask...
+                hi_mask_lo = self.data['hi'] == net_lo >> 64
+                hi_mask_hi = self.data['hi'] == net_hi >> 64
+                hi_mask = (self.data['hi'] > 0) & hi_mask_lo & hi_mask_hi
+                # This seems incorrect...
+                lo_mask = self.data['lo'] < (net_hi - net_lo)
+                return hi_mask & lo_mask
```

Comment:
Again, this maybe should be
```
    (self.data['lo'] & int(other.netmask) & lo_mask) == (int(other.network) & lo_mask) & (
         (self.data['hi'] & int(other.netmask)>>64) == (int(other.network)>>64)
```
(`lo_mask` is `0xffffffff`).

Reply:
Did you mean

```python
(self.data['lo'] & int(other.netmask) & lo_mask) == (int(other.network_address) & lo_mask) & (
                        (self.data['hi'] & int(other.netmask)>>64) == (int(other.network_address)>>64))
```

with `.network_address` instead of `network`? That faills the test for

```python
    s = ip.IPArray([u'2001:db8::1', u'2001:db9::1'])
    result = s.isin([u'2001:db8::0/96'])
```

It says both are in the network, while `'2001:db9::1'` should not be (I think).

I'll take a closer look at your approach though. This stuff is confusing for me :/